caikit/interfaces/ts/data_model/timeseries.py

```diff
@@ -0,0 +1,279 @@
+                from ..data_model.backends._spark_backends import (
+                    SparkMultiTimeSeriesBackend,
+                )
+
+                self._backend = SparkMultiTimeSeriesBackend(*args, **kwargs)
+
+    def __len__(self):
+        backend = getattr(self, "_backend", None)
+
+        if backend is None:
+            return len(self.as_pandas())
+
+        if HAVE_PYSPARK:
+            # Local
+            # pylint: disable=import-outside-toplevel
+            from ..data_model.backends._spark_backends import (
+                SparkMultiTimeSeriesBackend,
+            )
+
+        if isinstance(backend, PandasMultiTimeSeriesBackend):
+            return len(backend._df)
+        if HAVE_PYSPARK and isinstance(self._backend, SparkMultiTimeSeriesBackend):
+            return backend._pyspark_df.count()
+
+        error.log_raise(
+            "<COR75394521E>",
+            f"Unknown backend {type(backend)}",
+        )  # pragma: no cover
+
+    def _get_pd_df(self) -> Tuple[pd.DataFrame, Iterable[str], str, Iterable[str]]:
+        """Convert the data to a pandas DataFrame, efficiently if possible"""
+
+        # If there is a backend that knows how to do the conversion, use that
+        backend = getattr(self, "_backend", None)
+        if backend is not None and isinstance(backend, MultiTimeSeriesBackendBase):
+            log.debug("Using backend pandas conversion")
+            return backend.as_pandas()
+
+        error.value_check(
+            "<COR98388946E>",
+            self.timeseries is not None,
+            "Cannot create pandas data frame without any timeseries present",
+        )
+
+        error.value_check(
+            "<COR59303952E>",
+            self.id_labels is not None,
+            "Cannot create pandas data frame without any key labels present",
+        )
+
+        key_columns = self.id_labels
+        dfs = []
+        value_columns = None
+        timestamp_column = None
+        for ts in self.timeseries:  # pylint: disable=not-an-iterable
+            if value_columns is None:
+                value_columns = ts.value_labels
+                if ts.timestamp_label != "":
+                    timestamp_column = ts.timestamp_label
+            df = ts._get_pd_df()[0]
+
+            for i, key_col in enumerate(key_columns):
+                id_val = ts.ids.values[i]
+                df[key_col] = [id_val] * df.shape[0]
+            dfs.append(df)
+        ignore_index = True  # timestamp_column != ""
+        result = pd.concat(dfs, ignore_index=ignore_index)
+        self._backend = PandasMultiTimeSeriesBackend(
+            result,
+            key_column=key_columns,
+            timestamp_column=timestamp_column,
+            value_columns=value_columns,
+        )
+
+        return (
+            result,
+            key_columns,
+            timestamp_column,
+            value_columns,
+        )
+
+    def as_pandas(self, include_timestamps=None, is_multi=None) -> "pd.DataFrame":
```

Comment:
NIT: It would be good to have type hints on the args and have the args in the docstring